### PR TITLE
Add BetaBinomial Distribution

### PR DESCRIPTION
**Pull request by tblazina**
Add the BetaBinomial distribution

### CODE REVIEW

File: mcx/distributions/betabinomial.py
Comment (rlouf) [inline mcx/distributions/betabinomial.py:32]: The `loc` argument should be an attribute of the distribution. For now we'll only accept parametrization of the Beta distribution in terms of `a`, `b` and not `loc`, `scale` so you can remove `loc` altogether.

**Comment by rlouf** [comment]
Thank you for taking this up! I will also need exhaustive tests before merging this PR (take the [Beta distribution](https://github.com/rlouf/mcx/blob/master/tests/distributions/beta_test.py) as an example). What we want to test for:

- Sample shapes
- Logpdf shapes
- Make sure that `-inf` is returned by the logpdf when we're out of support.
- I think we can trust Demetri on testing the numerical correctness on the JAX side :)

**Comment by tblazina** [comment]
Of course! I plan to write tests, I just pushed this first commit to indicate that I'm working on it, I'll remove the `[WIP]` whenever I think it's actually ready for review, and I'd be happy to get feedback on any numerical correctness! 😃

**Comment by tblazina** [comment]
I see that the `jax.scipy.stats.betabinom` code was just added last week and I was wondering if it makes sense to hold off on this PR until jax 0.2.10 is released rather than duplicating/implementing it here?

**Comment by rlouf** [comment]
Yes, but JAX has a relatively short release cycle so we should be able
to merge quickly; we would have to bump the JAX version manually and
check that nothing breaks first.

**Comment by tblazina** [comment]
Ok sounds good, I'll continue writing the tests and everything and then when JAX 0.2.10  is released and we see it hasn't broke anything, it can be merged. Should the update of jax happen in this PR or separately, do you have a preference?

File: mcx/distributions/betabinomial.py
Comment (tblazina) [reply mcx/distributions/betabinomial.py:32]: Also another question here - I followed the current standard of calling the method `logpdf` even though this is a discrete distribution. What is the reason for not calling the method `logpmf` in the discrete case?

**Comment by rlouf** [comment]
Sounds good! Separate would be better.

**Comment by tblazina** [comment]
Ok great, I'll finish this up, let you know when its ready for review and tag that it needs to be merged only after the PR for updating JAX is merged. 👍

File: mcx/distributions/betabinomial.py
Comment (rlouf) [reply mcx/distributions/betabinomial.py:32]: To keep the API uniform. MCX appends `.logpdf` to each distribution when compiling the model's logpdf, so it's easier this way even though the terminology is not correct.

**Comment by rlouf** [comment]
Is it ready for review?

**Comment by tblazina** [comment]
Sure! Other than the commented out tests which look at the logpdf correctness which can be tested once jax 0.2.10 is released.  I more or less followed the tests in `beta_tests.py`, not sure if it would also make sense to add some 'edge case' tests as is done in the `binomial_tests.py` since the way it is implemented now is reusing the `_random_binomial` function defined in the `Binomial` distribution code.

File: mcx/distributions/betabinomial.py
Comment (rlouf) [inline mcx/distributions/betabinomial.py:8]:
```suggestion
from mcx.distributions.shapes import promote_shapes
```

File: mcx/distributions/betabinomial.py
Comment (rlouf) [inline mcx/distributions/betabinomial.py:20]:
```suggestion
        self.event_shape = ()
        n, a, b = promote_shapes(n, a, b)
```

File: mcx/distributions/betabinomial.py
Comment (rlouf) [inline mcx/distributions/betabinomial.py:24]:
```suggestion
		batch_shape = lax.broadcast_shapes(jnp.shape(n), jnp.shape(a), jnp.shape(b))
        self.batch_shape = batch_shape
        self.a = jnp.broadcast_shape(a, batch_shape)
        self.b = jnp.broadcast_shape(b, batch_shape)
        self.n = jnp.broadcast_shape(n, batch_shape)
```

File: mcx/distributions/betabinomial.py
Comment (rlouf) [inline mcx/distributions/betabinomial.py:1]:
```suggestion
from jax import lax
from jax import numpy as jnp
```

File: mcx/distributions/betabinomial.py
Comment (rlouf) [inline mcx/distributions/betabinomial.py:29]:
```suggestion
        n_max = jnp.max(self.n).item()
```

File: tests/distributions/beta_binomial_test.py
Comment (rlouf) [inline tests/distributions/beta_binomial_test.py:2]: Could you replace `np` by `jnp`?

File: tests/distributions/beta_binomial_test.py
Comment (rlouf) [inline tests/distributions/beta_binomial_test.py:65]: I think we should run these tests.

**Review by rlouf — CHANGES_REQUESTED** [review]
I suggested a few changes because I just merged a PR that changes the way shapes are handled in distributions. I also think tests on the logpdf should be run too. It should be ready to merge after that!

**Comment by tblazina** [comment]
Sorry had a sick kid at home. I'll make the changes and let you know when it's ready to merge!

**Comment by rlouf** [comment]
No problem, that's life and we're in no hurry! Is the kid feeling better?

**Comment by tblazina** [comment]
Ya she just started daycare so of course she got sick after 1 day and got us all sick. 🤷‍♂️ 
And my partner is applying for some professorships so has been very busy with application writing so ... child care has been falling on me until she sends those in. I'm full of excuses for why I have no time 😉 

In any case I found a bit of time this morning - I added the changes you requested, they all seem good to me. The tests fail because the PDF correctness tests reply on that `jax.scipy.stats.betabinomial` class which hasn't been released yet.

File: tests/distributions/beta_binomial_test.py
Comment (rlouf) [inline tests/distributions/beta_binomial_test.py:63]:
```suggestion
#
# LOGPDF CORRECTNESS
#
```

Nitpicking :)

File: tests/distributions/beta_binomial_test.py
Comment (rlouf) [inline tests/distributions/beta_binomial_test.py:123]: Even when we update to the newest version of JAX these tests will fail with the new shape management. A sample shape `()` will return a scalar, etc. Now shapes are promoted for broadcasting at initialization of the distributions, which makes a lot more sense.

**Comment by rlouf** [comment]
I just bumped JAX's version to 0.2.10, you can rebase to see if the tests pass.

**Comment by tblazina** [comment]
Ok , updated the shape tests and with jax 0.2.10 the tests pass 👍 
Unless yo have any other inputs, I think it's ready to merge.

File: tests/distributions/beta_binomial_test.py
Comment (tblazina) [reply tests/distributions/beta_binomial_test.py]: no worries - i'm the same way ;)

**Comment by rlouf** [comment]
Looks great to me! Thank you for taking the time to add this to the library!🙂

**Pull request merged**
